python/src/iceberg/table/schema.py

```diff
@@ -0,0 +1,346 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+import sys
+from typing import Iterable, Tuple
+
+if sys.version_info >= (3, 8):  # pragma: no cover
+    from typing import Protocol
+else:  # pragma: no cover
+    from typing_extensions import Protocol  # type: ignore
+
+if sys.version_info >= (3, 8):  # pragma: no cover
+    from functools import singledispatchmethod
+else:  # pragma: no cover
+    from singledispatch import singledispatchmethod  # type: ignore
+
+from iceberg.types import IcebergType, ListType, MapType, NestedField, StructType
+
+
+class SchemaVisitor(Protocol):
+    def visit(self, node) -> None:  # pragma: no cover
```

Comment:
Question: is this returning `None` and then using the `result` field instead because the returned dictionaries are somewhat different?

In either case, would it make sense to add a doc string to this Protocol, indicating that the `result` is used? Not sure if a good way to phrase it but having `None` as the return type makes me feel like it’s going to mutate `node` or some internal state (which I don’t believe they do).

Reply:
That's right with this current design the visitor doesn't manipulate `node` but it accumulates state as it visits different parts of the schema. I'm wondering if it would be too strict to require concrete visitors to store the state in an attribute specifically called `result` but if that's ok then I can actually make this an ABC class which would enforce it. (A child class would still be allowed to have arguments in addition to `node` if needed):
```py
from abc import ABC, abstractmethod

class SchemaVisitor(ABC):

    @abstractmethod
    def visit(self, node) -> None:
        ...

    @property
    @abstractmethod
    def result(self):
        ...
```
Let me know what you think.